### PR TITLE
Clean up and document ModelMatrix logic

### PR DESCRIPTION
**Pull request by dmbates**
Cf. https://github.com/JuliaStats/DataFrames.jl/issues/1012.

### CODE REVIEW

File: src/statsmodels/formula.jl [hunk 1]
```diff
@@ -246,78 +235,145 @@ end
 ModelFrame(f::Formula, d::AbstractDataFrame) = ModelFrame(Terms(f), d)
 ModelFrame(ex::Expr, d::AbstractDataFrame) = ModelFrame(Formula(ex), d)
 
+asMatrix(a::AbstractMatrix) = a
+asMatrix(v::AbstractVector) = reshape(v, (length(v), 1))
+
+"""
+    StatsBase.model_response(mf::ModelFrame)
+Extract the response column, if present.  `DataVector` or
```
Comment (ararslan) [inline src/statsmodels/formula.jl]: Need a line break after the indented line for the markdown to render properly

File: src/statsmodels/formula.jl [hunk 2]
```diff
@@ -246,78 +235,145 @@ end
...
+    end
 end
 
+"""
+    contr_treatment(n::Integer, contrasts::Bool, sparse::Bool, base::Integer)
+Create a sparse or dense identity of size `n`.  Return the identity if `contrasts`
+is false.  Otherwise drop the `base` column.
+"""
 function contr_treatment(n::Integer, contrasts::Bool, sparse::Bool, base::Integer)
     if n < 2 error("not enought degrees of freedom to define contrasts") end
```
Comment (ararslan) [inline src/statsmodels/formula.jl]: It'd be nice to use explicitly typed exceptions, e.g.
```julia
n < 2 && throw(ArgumentError("Not enough degrees of freedom to define contrasts"))
```
Also you have a trailing "t" on "enough."

File: src/statsmodels/formula.jl [hunk 3]
```diff
@@ -246,78 +235,145 @@ end
...
     contr = sparse ? speye(n) : eye(n) .== 1.
-    if !contrasts return contr end
-    if !(1 <= base <= n) error("base = $base is not allowed for n = $n") end
-    contr[:,vcat(1:(base-1),(base+1):end)]
+    if !contrasts
+        contr
+    elseif !(1 <= base <= n)
+        error("base = $base is not allowed for n = $n")
```
Comment (ararslan) [inline src/statsmodels/formula.jl]: Could use `ArgumentError` here too

File: src/statsmodels/formula.jl [hunk 4]
```diff
@@ -246,78 +235,145 @@ end
 ModelFrame(f::Formula, d::AbstractDataFrame) = ModelFrame(Terms(f), d)
 ModelFrame(ex::Expr, d::AbstractDataFrame) = ModelFrame(Formula(ex), d)
 
+asMatrix(a::AbstractMatrix) = a
```
Comment (ararslan) [inline src/statsmodels/formula.jl]: This doesn't follow the Julia function naming convention of all lowercase or lower snake case.

File: src/statsmodels/formula.jl [hunk 5]
```diff
@@ -246,78 +235,145 @@ end
...
+cols(v::Vector) = asMatrix(convert(Vector{Float64}, v))
+
+"""
+    expandcols(trm::Vector)
+Create pairwise products of columns from a vector of matrices
+"""
 function expandcols(trm::Vector)
     if length(trm) == 1
-        return convert(Array{Float64}, trm[1])
+        asMatrix(convert(Array{Float64}, trm[1]))
```
Comment (ararslan) [inline src/statsmodels/formula.jl]: `trm` is already a vector, why convert to an array?

File: src/statsmodels/formula.jl [hunk 6]
```diff
@@ -246,78 +235,145 @@ end
...
     end
-    n
 end
 
+"""
+    dropResponse(trms::Terms)
+Drop the response term, `trms.eterms[1]` and the first row and column
+of `trms.factors` if `trms.response` is true.
+"""
+dropResponse(trms::Terms) = !trms.response ? trms :
```
Comment (ararslan) [inline src/statsmodels/formula.jl]: `dropresponse` or `drop_response`

Comment (ararslan) [inline src/statsmodels/formula.jl]: Why not just modify `remove_response`?

File: src/statsmodels/formula.jl [hunk 7]
```diff
@@ -246,78 +235,145 @@ end
...
-    n = 1
-    for x in trm
-        n *= size(x, 2)
+"""
+    dropRanefTerms(trms::Terms)
+Expressions of the form `(a|b)` are "random-effects" terms and are not
+incorporated in the ModelMatrix.  This function checks for such terms and,
+if any are present, drops them from the `Terms` object.
+"""
+function dropRanefTerms(trms::Terms)
```
Comment (ararslan) [inline src/statsmodels/formula.jl]: Besides the camel case, this function name is kind of confusing. Even `droprandomeffects` would be better IMO.

**Comment by ararslan** [comment]
I dunno why Travis is mad about `view`. If it's been a while you might want to rebase from master.

File: src/statsmodels/formula.jl (hunk 1 above)
Comment (dmbates) [inline src/statsmodels/formula.jl]: Actually you don't need a line break.  A line with a 4 space indent followed by a line with no indent causes a new markdown block.

File: src/statsmodels/formula.jl (hunk 2 above)
Comment (dmbates) [inline src/statsmodels/formula.jl]: Thanks for catching that.

By the way, I think the
```
    if condition
        throw(ArgumentError(...))
    end
```
is preferred to the more compact
```
    condition && throw(...
```
for checking code coverage.  If the check and error throwing are on the same line you can't tell if the error is ever thrown in the tests.

File: src/statsmodels/formula.jl (hunk 4 above)
Comment (dmbates) [inline src/statsmodels/formula.jl]: Thanks.

File: src/statsmodels/formula.jl (hunk 6 above)
Comment (dmbates) [inline src/statsmodels/formula.jl]: `CategoricalArrays` has also used the `drop` verb in `droplevels!` so I will use `dropresponse!`.  It should have the `!` in the name.

File: src/statsmodels/formula.jl (hunk 1 above)
Comment (ararslan) [inline src/statsmodels/formula.jl]: Oh neat I didn't know that! Thanks for letting me know. :)

**Comment by ararslan** [comment]
This is great work. Thanks so much for doing this, Doug! You rock! 🎸 :metal: 👏

File: src/statsmodels/formula.jl (hunk 2 above)
Comment (ararslan) [inline src/statsmodels/formula.jl]: Oh that's good to know, thanks.

**Comment by kleinschmidt** [comment]
This is great, thanks for doing it Doug. I think the only place where there's substantive overlap with #870 is in the ModelMatrix code itself. In order to construct columns for categorical variables, it's necessary to check whether any lower-order terms are present (e.g., an intercept column for a main effect `A`, or a main effect `A` for an interaction `A&B`). But I think that should be pretty simple to rebase on top of this PR.

**Comment by dmbates** [comment]
@ararslan Thanks.  I appreciate your careful reading of the changes.

@kleinschmidt I like your approach to allowing more general contrasts.  I can look into merging those changes in the next couple of days.  It would be worthwhile refactoring Terms in general.  I unabashedly patterned it after the R terms object but I think it would be much easier to do better than that now that I understand Julia a bit better.

The Appveyor failure seems unrelated to this PR.  It failed on only one of the 3 configurations and that was in the tests for RDA.

I merge this tomorrow unless someone else gets there ahead of me.

**Comment by kleinschmidt** [comment]
You might also be interested in looking at https://github.com/kleinschmidt/Formulas.jl

I played around a bit with a representation of terms that uses type parameters and multiple dispatch. It substantially simplifies the rewrite rules. The code is pretty ugly as it stands (lots of splatting) but I think it's a clearer and more expressive approach than the R-style one.

File: src/statsmodels/formula.jl (hunk 1 above)
Comment (nalimilan) [inline src/statsmodels/formula.jl]: Still, please add a line break for consistency with other docstrings, in particular in Julia Base.

File: src/statsmodels/formula.jl (hunk 6 above)
Comment (kleinschmidt) [inline src/statsmodels/formula.jl]: I find the `!` confusing. I thought it was idiomatically used for things that modified their first argument, but `dropresponse!` doesn't (it returns a new `Terms` instance).

**Comment by nalimilan** [comment]
@dmbates You merged without fixing some naming issues highlighted during the review. Could you fix them in another PR?

**Comment by nalimilan** [comment]
Nevermind, got confused while viewing an older version of the diff.

**Comment by tkelman** [comment]
This was just released in a tagged version yesterday, and I think it caused problems in StatsBase, FixedEffectModels, and MixedModels
http://pkg.julialang.org/logs/StatsBase_0.4.log
http://pkg.julialang.org/logs/FixedEffectModels_0.4.log
http://pkg.julialang.org/logs/MixedModels_0.4.log

**Comment by ararslan** [comment]
Actually I think it was #870 that caused the issues, at least the one in StatsBase.

**Comment by tkelman** [comment]
Whoops. I'll leave it to people who know what they're talking about. Do we want to do the same thing we did with Distributions, and make a new less-breaking patch tag off a branch from before either or both of these large recent changes and with just the deprecation fixes cherry-picked?

**Comment by ararslan** [comment]
In the interest of fire extinguishing, that's probably a good idea. (If we're quick we could even get to it before PkgEvaluator does.)

**Comment by tkelman** [comment]
We've got an hour before it starts, though I can tweak my cronjob and push it back a bit if that's too tight.

**Comment by ararslan** [comment]
Shouldn't be necessary. https://github.com/JuliaLang/METADATA.jl/pull/5893

**Comment by matthieugomez** [comment]
This commit introduced a pretty bad error in my package (i.e. a function in my package would return a wrong numerical result). It was caught by the tests in my package as noticed by @tkelman , althought it was hidden by another test failure due to #1035.

The issue is that ModelMatrix becomes view on the underlying dataframe in the case where is only one vector
```julia
using RDatasets
df = dataset("plm", "Cigar")
mf = ModelFrame(Sales ~ 0 + Price, df)
X = ModelMatrix(mf).m
fill!(X, 0.0)
mf.df[:Price]
```

**Comment by nalimilan** [comment]
@matthieugomez So this could simply be fixed by ensuring we copy the column? Would you make a PR?

**Pull request merged**